src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -1271,38 +1276,81 @@ JoinRangeTableEntry(JoinExpr *joinExpr, List *dependentJobList, List *rangeTable
 	rangeTableEntry->eref->colnames = joinedColumnNames;
 	rangeTableEntry->joinaliasvars = joinedColumnVars;
 
-	SetJoinRelatedColumnsCompat(rangeTableEntry,
-								leftColumnNames, rightColumnNames, leftColumnVars,
+	SetJoinRelatedColumnsCompat(rangeTableEntry, leftRelId, rightRelId, leftColumnVars,
 								rightColumnVars);
 
 	return rangeTableEntry;
 }
 
 
 static void
-SetJoinRelatedColumnsCompat(RangeTblEntry *rangeTableEntry,
-							List *leftColumnNames, List *rightColumnNames,
+SetJoinRelatedColumnsCompat(RangeTblEntry *rangeTableEntry, Oid leftRelId, Oid rightRelId,
 							List *leftColumnVars, List *rightColumnVars)
 {
 	#if PG_VERSION_NUM >= PG_VERSION_13
 
 	/* We don't have any merged columns so set it to 0 */
 	rangeTableEntry->joinmergedcols = 0;
-	int numvars = list_length(leftColumnVars);
-	for (int varId = 1; varId <= numvars; varId++)
+
+	if (OidIsValid(leftRelId))
+	{
+		rangeTableEntry->joinleftcols = GetColumnOriginalIndexes(leftRelId);
+	}
+	else
+	{
+		int numvars = list_length(leftColumnVars);
+		for (int varId = 1; varId <= numvars; varId++)
+		{
+			rangeTableEntry->joinleftcols = lappend_int(rangeTableEntry->joinleftcols,
+														varId);
+		}
+	}
+
+	if (OidIsValid(rightRelId))
 	{
-		rangeTableEntry->joinleftcols = lappend_int(rangeTableEntry->joinleftcols, varId);
+		rangeTableEntry->joinrightcols = GetColumnOriginalIndexes(rightRelId);
 	}
-	numvars = list_length(rightColumnVars);
-	for (int varId = 1; varId <= numvars; varId++)
+	else
 	{
-		rangeTableEntry->joinrightcols = lappend_int(rangeTableEntry->joinrightcols,
-													 varId);
+		int numvars = list_length(rightColumnVars);
+		for (int varId = 1; varId <= numvars; varId++)
+		{
+			rangeTableEntry->joinrightcols = lappend_int(rangeTableEntry->joinrightcols,
+														 varId);
+		}
 	}
+
 	#endif
 }
 
 
+#if PG_VERSION_NUM >= PG_VERSION_13
+
+/*
+ * GetColumnOriginalIndexes gets the original indexes of columns by taking column drops into account.
+ */
+static List *
+GetColumnOriginalIndexes(Oid relationId)
+{
+	List *originalIndexes = NIL;
+	Relation relation = table_open(relationId, NoLock);
```

Comment:
Why don't we acquire any locks ?

I guess this is because we assume caller already took locks but we can acquire what postgres would acquire in that case (probably `AccessShareLock`) here for further usages of this function.

Reply:
+ I think we shouldn't release the lock when closing the relation unless we have a specific reason for doing it